dpnp/dpnp_utils/dpnp_utils_pad.py

```diff
@@ -0,0 +1,784 @@
+    axis : int
+        The currently padded dimension assuming that `axis` is padded before
+        `axis` + 1.
+
+    Returns
+    -------
+    roi : ndarray
+        The region of interest of the original `array`.
+
+    """
+
+    axis += 1
+    sl = (slice(None),) * axis + original_area_slice[axis:]
+    return array[sl]
+
+
+def dpnp_pad(array, pad_width, mode="constant", **kwargs):
+    """Pad an array."""
+
+    if isinstance(pad_width, int):
+        if pad_width < 0:
+            raise ValueError("index can't contain negative values")
+        pad_width = ((pad_width, pad_width),) * array.ndim
+    else:
+        if dpnp.is_supported_array_type(pad_width):
+            pad_width = dpnp.asnumpy(pad_width)
+        else:
+            pad_width = numpy.asarray(pad_width)
+
+        if not pad_width.dtype.kind == "i":
+            raise TypeError("`pad_width` must be of integral type.")
+
+        # Broadcast to shape (array.ndim, 2)
+        pad_width = _as_pairs(pad_width, array.ndim, as_index=True)
+
+    if callable(mode):
+        function = mode
+        # Create a new zero padded array
+        padded, _ = _pad_simple(array, pad_width, fill_value=0)
+        # And apply along each axis
+
+        for axis in range(padded.ndim):
+            # Iterate using ndindex as in apply_along_axis, but assuming that
+            # function operates inplace on the padded array.
+
+            # view with the iteration axis at the end
+            view = dpnp.moveaxis(padded, axis, -1)
+
+            # compute indices for the iteration axes, and append a trailing
+            # ellipsis to prevent 0d arrays decaying to scalars
+            inds = numpy.ndindex(view.shape[:-1])
+            inds = (ind + (Ellipsis,) for ind in inds)
+            for ind in inds:
+                function(view[ind], pad_width[axis], axis, kwargs)
+
+        return padded
+
+    # Make sure that no unsupported keywords were passed for the current mode
+    allowed_kwargs = {
+        "empty": [],
+        "edge": [],
+        "wrap": [],
+        "constant": ["constant_values"],
+        "linear_ramp": ["end_values"],
+        "maximum": ["stat_length"],
+        "mean": ["stat_length"],
+        # "median": ["stat_length"],  # TODO: dpnp.median is not implemented
+        "minimum": ["stat_length"],
+        "reflect": ["reflect_type"],
+        "symmetric": ["reflect_type"],
+    }
+
+    try:
+        unsupported_kwargs = set(kwargs) - set(allowed_kwargs[mode])
+    except KeyError:
+        raise ValueError(f"mode '{mode}' is not supported") from None
+    if unsupported_kwargs:
+        raise ValueError(
+            f"unsupported keyword arguments for mode '{mode}': {unsupported_kwargs}"
+        )
+
+    if mode == "constant":
+        values = kwargs.get("constant_values", 0)
+        if (
+            dpnp.isscalar(values)
+            and values == 0
+            and (array.ndim == 1 or array.size < 4e6)
```

Comment:
We need to measure and to check the size of input array where to use the fast path. It might differ from cupy I guess.

Reply:
It depends on the performance of `fill` method. I am waiting for https://github.com/IntelPython/dpnp/pull/2055 to be merged first.

Reply:
Updated the limiting size to 3*10^7 based on the following timing obtained on Data Center GPU Max 1100.

```python
import dpnp
a=dpnp.ones((10000, 1000))
%timeit dpnp.pad(a, [2,3], "constant")
```

| Size       | Fast Path          | Slow Path           |
|------------|---------------------|---------------------|
| 1*10^7     | 380 μs ± 12.8 μs    | 736 μs ± 2.45 μs    |
| 2*10^7     | 820 μs ± 1.27 μs    | 1.04 ms ± 1.73 μs   |
| 3*10^7     | 1.23 ms ± 2.82 μs   | 1.38 ms ± 2.55 μs   |
| 4*10^7     | 1.63 ms ± 3.57 μs   | 1.54 ms ± 49.9 μs   |
| 5*10^7     | 2.03 ms ± 2.74 μs   | 1.93 ms ± 111 μs    |